ct/l1: Fix object builder uninitialized state

wdberkeley opened the pull request:
The object builder was tracking partition state with _current_tidp, which was uninitialized. This caused a garbage partitions to be added to the object index when end_partition() was called before any partition was started. I changed _current_tidp to std::optional to fix it.

I found this bug rebasing the reconciler object building change on the conversion of object builder to use topic id partition. That change contains a regression test for this issue.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [X] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj commented on src/v/cloud_topics/level_one/common/object.cc:
```diff
@@ -411,7 +413,7 @@ class object_builder_impl final : public object_builder {
     size_t _offset = 0;
     ss::output_stream<char> _output;
     footer _index;
-    model::topic_id_partition _current_tidp;
```
Ah default initialization strikes again. I would have thought UUIDs are initialized to all zeros but that's not the case and there are too many layers for clang tidy to understand and warn me.

This also fixes it and also adds a test:

```diff
diff --git a/src/v/cloud_topics/level_one/common/object.cc b/src/v/cloud_topics/level_one/common/object.cc
index df2f7418f9..b312f08e45 100644
--- a/src/v/cloud_topics/level_one/common/object.cc
+++ b/src/v/cloud_topics/level_one/common/object.cc
@@ -411,7 +411,7 @@ private:
     size_t _offset = 0;
     ss::output_stream<char> _output;
     footer _index;
-    model::topic_id_partition _current_tidp;
+    model::topic_id_partition _current_tidp{};
     footer::partition _current_partition;
     options _opts;
 };
diff --git a/src/v/cloud_topics/level_one/common/tests/object_test.cc b/src/v/cloud_topics/level_one/common/tests/object_test.cc
index 811bc6616c..d8309985b9 100644
--- a/src/v/cloud_topics/level_one/common/tests/object_test.cc
+++ b/src/v/cloud_topics/level_one/common/tests/object_test.cc
@@ -498,6 +498,7 @@ TEST(L1Objects, FullScan) {
       },
     };
     auto [info, object] = make_object(specs_by_tidp);
+    EXPECT_EQ(info.index.partitions.size(), 2);
     EXPECT_EQ(info.size_bytes, object.size_bytes());
     std::variant<footer, size_t> read_footer_result;
     ASSERT_NO_THROW(
```

dotnwat replied:
shouldn't we do this? it's generally useful to have predictable init for our named_types.

```
diff --git a/src/v/utils/uuid.h b/src/v/utils/uuid.h
index effb5a4723..b442f3dfe7 100644
--- a/src/v/utils/uuid.h
+++ b/src/v/utils/uuid.h
@@ -64,7 +64,7 @@ private:
     explicit uuid_t(const underlying_t& uuid)
       : _uuid(uuid) {}

-    underlying_t _uuid;
+    underlying_t _uuid{};
 };

 bool operator<(const uuid_t& l, const uuid_t& r);
```

rockwotj replied:
yeah I made the same patch, let me send it out

rockwotj reviewed (commented):
Nice catch! Let's please add a test. I added a simple test in my comment.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--63f418b8-3be9-40e9-81d3-2362f58be5c3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71874">build#71874</a></summary>


| test_class                       | test_method                        | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:---------------------------------|:-----------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest            | test_migrated_topic_data_integrity | {"include_groups": true, "params": {"cancellation": {"dir": "in", "stage": "preparing"}, "use_alias": false}, "transfer_leadership": true} | integration | https://buildkite.com/redpanda/redpanda/builds/71874#01992c8a-635d-40c9-9335-dc8d19b19ec4 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss         | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/71874#01992c83-9c9c-4e84-9861-9212b90d5378 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

wdberkeley committed:
ct/l1: Fix object builder uninitialized state

The object builder was tracking partition state with _current_tidp,
which was uninitialized. This caused a garbage partitions to be added
to the object index when end_partition() was called before any partition
was started. I changed _current_tidp to std::optional to fix it.
